state-commitments/optimized/inclusion_proof.go

```diff
@@ -0,0 +1,113 @@
+package optimized
+
+import (
+	"errors"
+	"math"
+
+	"github.com/ethereum/go-ethereum/common"
+)
+
+// Computes the Merkle proof for a leaf at a given index.
+// It uses the last leaf to pad the tree up to the 'virtual' size if needed.
+func (h *HistoryCommitter) computeMerkleProof(leafIndex uint64, leaves []common.Hash, virtual uint64) ([]common.Hash, error) {
+	if len(leaves) == 0 {
+		return nil, nil
+	}
+	ok, err := isPowTwo(virtual)
+	if err != nil {
+		return nil, err
+	}
+	if !ok {
+		return nil, errors.New("virtual size must be a power of 2")
+	}
+	if leafIndex >= uint64(len(leaves)) {
+		return nil, errors.New("leaf index out of bounds")
+	}
+	if virtual < uint64(len(leaves)) {
+		return nil, errors.New("virtual size must be greater than or equal to the number of leaves")
+	}
+	numRealLeaves := uint64(len(leaves))
+	lastLeaf, err := h.hash(leaves[numRealLeaves-1][:])
+	if err != nil {
+		return nil, err
+	}
+	depth := int(math.Ceil(math.Log2(float64(virtual))))
+
+	// Precompute virtual hashes
+	virtualHashes, err := h.precomputeRepeatedHashes(&lastLeaf, depth)
+	if err != nil {
+		return nil, err
+	}
+	var proof []common.Hash
+	for level := 0; level < depth; level++ {
+		nodeIndex := leafIndex >> level
+		siblingHash, exists, err := h.computeSiblingHash(nodeIndex, uint64(level), numRealLeaves, virtual, leaves, virtualHashes)
+		if err != nil {
+			return nil, err
+		}
+		if exists {
+			proof = append(proof, siblingHash)
+		}
+	}
+	return proof, nil
+}
+
+// Computes the hash of a node's sibling at a given index and level.
+func (h *HistoryCommitter) computeSiblingHash(
+	nodeIndex uint64,
+	level uint64,
+	N uint64,
```

Comment:
I think `N` is `numRealNodes`, which might be a bit clearer (and starts with a lowercase letter)